src/main/java/eu/arrowhead/deviceqosevaluator/service/QualityEvaluationService.java

```diff
@@ -0,0 +1,175 @@
+
+	@Autowired
+	private QualitiyEvaluationValidation validator;
+
+	@Autowired
+	private StatisticsEngine statEngine;
+
+	@Autowired
+	private ObjectMapper mapper;
+
+	private final Logger logger = LogManager.getLogger(this.getClass());
+
+	//=================================================================================================
+	// methods
+
+	//-------------------------------------------------------------------------------------------------
+	public QoSEvaluationFilterResponseDTO filter(final QoSEvaluationRequestDTO dto, final String origin) {
+		logger.debug("filter started");
+
+		try {
+			final List<String> systems = dto == null ? null : dto.providers();
+			final QoSDeviceDataEvaluationConfigDTO config = dto == null || dto.configuration() == null ? null : mapper.readValue(Utilities.toJson(dto.configuration()), QoSDeviceDataEvaluationConfigDTO.class);
+
+			final Pair<List<String>, QoSDeviceDataEvaluationConfigDTO> normalized = validator.validateAndNormalizeQoSEvaluationRequest(systems, config, true, origin);
+
+			final List<SystemEvalModel> evaluated = evaluateRequest(normalized.getFirst(), normalized.getSecond());
+			final List<String> passedProviders = new ArrayList<>();
+			final List<String> droppedProviders = new ArrayList<>();
+			final Map<String, List<String>> warnings = new HashMap<>();
+
+			for (final SystemEvalModel sysResult : evaluated) {
+				if (sysResult.getScore() <= config.threshold()) {
+					passedProviders.add(sysResult.getName());
+				} else {
+					droppedProviders.add(sysResult.getName());
+				}
+
+				if (!Utilities.isEmpty(sysResult.getNoStat())) {
+					warnings.put(sysResult.getName(), sysResult.getNoStat().stream().map(item -> item.name()).toList());
+				}
+			}
+
+			return new QoSEvaluationFilterResponseDTO(passedProviders, droppedProviders, warnings);
+
+		} catch (final JsonProcessingException ex) {
+			throw new InvalidParameterException("Invalid configuration payload", origin);
+		}
+	}
+
+	//-------------------------------------------------------------------------------------------------
+	public QoSEvaluationSortResponseDTO sort(final QoSEvaluationRequestDTO dto, final String origin) {
+		logger.debug("sort started");
+
+		try {
+			final List<String> systems = dto == null ? null : dto.providers();
+			final QoSDeviceDataEvaluationConfigDTO config = dto == null || dto.configuration() == null ? null : mapper.readValue(Utilities.toJson(dto.configuration()), QoSDeviceDataEvaluationConfigDTO.class);
+
+			final Pair<List<String>, QoSDeviceDataEvaluationConfigDTO> normalized = validator.validateAndNormalizeQoSEvaluationRequest(systems, config, false, origin);
+
+			final List<SystemEvalModel> evaluated = evaluateRequest(normalized.getFirst(), normalized.getSecond());
+			final List<String> sortedProviders = new ArrayList<>();
+			final Map<String, List<String>> warnings = new HashMap<>();
+
+			evaluated.sort(Comparator.comparingDouble(SystemEvalModel::getScore)); //ascending
```

Comment:
I think this should go to inside of evaluateRequest, because even in case of filtering, it is good if the first one in the passed provider is the best.